lib/auth/tls_test.go

```diff
@@ -1357,6 +1357,99 @@ func (s *TLSSuite) TestGetCertAuthority(c *check.C) {
 	fixtures.ExpectAccessDenied(c, err)
 }
 
+func (s *TLSSuite) TestAccessRequest(c *check.C) {
+	priv, pub, err := s.server.Auth().GenerateKeyPair("")
+	c.Assert(err, check.IsNil)
+
+	// make sure we can parse the private and public key
+	privateKey, err := ssh.ParseRawPrivateKey(priv)
+	c.Assert(err, check.IsNil)
+
+	_, err = tlsca.MarshalPublicKeyFromPrivateKeyPEM(privateKey)
+	c.Assert(err, check.IsNil)
+
+	_, _, _, _, err = ssh.ParseAuthorizedKey(pub)
+	c.Assert(err, check.IsNil)
+
+	user := "user1"
+	role := "some-role"
+	_, err = CreateUserRoleAndRequestable(s.server.Auth(), user, role)
+	c.Assert(err, check.IsNil)
+
+	testUser := TestUser(user)
+	testUser.TTL = time.Hour
+	userClient, err := s.server.NewClient(testUser)
+	c.Assert(err, check.IsNil)
+
+	req, err := services.NewAccessRequest(user, role)
+	c.Assert(err, check.IsNil)
+
+	c.Assert(userClient.CreateAccessRequest(req), check.IsNil)
+
+	// sanity check; ensure that roles for which no `allow` directive
+	// exists cannot be requested.
+	badReq, err := services.NewAccessRequest(user, "some-fake-role")
+	c.Assert(err, check.IsNil)
+	c.Assert(userClient.CreateAccessRequest(badReq), check.NotNil)
+
+	// generateCerts executes a GenerateUserCerts request, optionally applying
+	// one or more access-requests to the certificate.
+	generateCerts := func(reqIDs ...string) (*proto.Certs, error) {
+		return userClient.GenerateUserCerts(context.TODO(), proto.UserCertsRequest{
+			PublicKey:      pub,
+			Username:       user,
+			Expires:        time.Now().Add(time.Hour).UTC(),
+			Format:         teleport.CertificateFormatStandard,
+			AccessRequests: reqIDs,
+		})
+	}
+
+	// certContainsRole checks if a PEM encoded TLS cert contains the
+	// specified role.
+	certContainsRole := func(cert []byte, role string) bool {
+		tlsCert, err := tlsca.ParseCertificatePEM(cert)
+		c.Assert(err, check.IsNil)
+		identity, err := tlsca.FromSubject(tlsCert.Subject, tlsCert.NotAfter)
+		c.Assert(err, check.IsNil)
+		return utils.SliceContainsStr(identity.Groups, role)
+	}
+
+	// sanity check; ensure that role is not held if no request is applied.
+	userCerts, err := generateCerts()
+	c.Assert(err, check.IsNil)
+	if certContainsRole(userCerts.TLS, role) {
+		c.Errorf("unexpected role %s", role)
+	}
+
+	// attempt to apply request in PENDING state (should fail)
+	_, err = generateCerts(req.GetName())
+	c.Assert(err, check.NotNil)
+
+	// verify that user does not have the ability to approve their own request (not a special case, this
+	// user just wasn't created with the necessary roles for request management).
+	c.Assert(userClient.SetAccessRequestState(req.GetName(), services.RequestState_APPROVED), check.NotNil)
```

Comment:
In this case does the user automatically obtain that role?  This could be the case that a user has an Admin , but has a Super Admin role that they use to modify SSO connectors, and by requesting it creates a better audit for 'super admin' activities.

Reply:
@benarent there is no magic, and I'm hesitant to add any.  That being said, the permissions are set up such that a user with the correct permissions could technically create an `AccessRequest` resource which was *already* in the `APPROVED` state.  No command to do so is exposed by `tsh` at the moment, but its possible.

Reply:
Thanks for the clarification, I'm going to resolve this for now.